File: tests/rng/rng_test.c
#include "../new_src_c/rng.h"
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>

int main(int argc, char **argv) {
    prng_state rs;
    int user_seed = 12345;
    if (argc >= 2) {
        user_seed = atoi(argv[1]);
    }

    uint32_t p[8];
    for (int i = 0; i < 8; i++) {
        p[i] = user_seed + i;
    }

    uint64_t s_packed[4];
    s_packed[0] = ((uint64_t)p[6] << 32) | (uint64_t)p[7];
    s_packed[1] = ((uint64_t)p[4] << 32) | (uint64_t)p[5];
    s_packed[2] = ((uint64_t)p[2] << 32) | (uint64_t)p[3];
    s_packed[3] = ((uint64_t)p[0] << 32) | (uint64_t)p[1];

    manual_seed(&rs, s_packed);

    printf("Uniforms:\n");
    for (int i = 0; i < 10; i++) {
        printf("%20.16f\n", rand_uniform(&rs, 0.0, 1.0));
    }

    printf("Normals:\n");
    for (int i = 0; i < 10; i++) {
        printf("%20.16f\n", rand_normal(&rs, 0.0, 1.0));
    }

    printf("Gammas (shape=2.0, scale=1.0):\n");
    for (int i = 0; i < 10; i++) {
        printf("%20.16f\n", rand_gamma(&rs, 2.0, 1.0));
    }

    return 0;
}

File: tests/rng/rng_test.f90
program rng_test
    use mod_defs, only: dp
    use mod_random
    implicit none

    integer :: i, n
    integer, dimension(:), allocatable :: seed
    character(len=20) :: arg
    integer :: user_seed

    ! Initialize the seed from the command line
    if (command_argument_count() >= 1) then
        call get_command_argument(1, arg)
        read(arg, *) user_seed
    else
        user_seed = 12345
    end if

    call random_seed(size=n)
    allocate(seed(n))
    ! Seeding logic matching Fortran 13 / libgfortran default
    ! Let's just use user_seed for all elements or something predictable
    seed = [(user_seed + i - 1, i = 1, n)]
    call random_seed(put=seed)

    print *, "Uniforms:"
    do i = 1, 10
        print "(F20.16)", rand_uniform(0.0_dp, 1.0_dp)
    end do

    print *, "Normals:"
    do i = 1, 10
        print "(F20.16)", rand_normal(0.0_dp, 1.0_dp)
    end do

    print *, "Gammas (shape=2.0, scale=1.0):"
    do i = 1, 10
        print "(F20.16)", rand_gamma(2.0_dp, 1.0_dp)
    end do

end program rng_test
